File: Codes/3297-count-substrings-that-can-be-rearranged-to-contain-a-string-i.py
'''
Author: LetMeFly
Date: 2025-01-09 12:39:58
LastEditors: LetMeFly.xyz
LastEditTime: 2025-01-09 12:44:30
'''
from collections import Counter, defaultdict


class Solution:
    def ok(self, cnt1: defaultdict) -> bool:
        for k, v in self.cnt2.items():
            if cnt1[k] < v:
                return False
        return True
    
    def validSubstringCount(self, word1: str, word2: str) -> int:
        self.cnt2 = Counter(word2)
        cnt1 = defaultdict(int)
        ans = l = r = 0
        while l < len(word1):
            if l:
                cnt1[word1[l - 1]] -= 1
            while not self.ok(cnt1):
                if r == len(word1):
                    return ans
                cnt1[word1[r]] += 1
                r += 1
            ans += len(word1) - r + 1
            l += 1
        return ans

File: Codes/3297-count-substrings-that-can-be-rearranged-to-contain-a-string-i.java
/*
 * @Author: LetMeFly
 * @Date: 2025-01-09 12:46:14
 * @LastEditors: LetMeFly.xyz
 * @LastEditTime: 2025-01-09 12:51:13
 */
class Solution {
    private boolean ok(int[] a, int[] b) {
        for (int i = 0; i < 26; i++) {
            if (a[i] < b[i]) {
                return false;
            }
        }
        return true;
    }

    public long validSubstringCount(String word1, String word2) {
        int[] cnt1 = new int[26], cnt2 = new int[26];
        for (char c : word2.toCharArray()) {
            cnt2[c - 'a']++;
        }
        long ans = 0;
        for (int l = 0, r = 0; l < word1.length(); l++) {
            if (l > 0) {
                cnt1[word1.charAt(l - 1) - 'a']--;
            }
            while (!ok(cnt1, cnt2)) {
                if (r == word1.length()) {
                    return ans;
                }
                cnt1[word1.charAt(r++) - 'a']++;
            }
            ans += word1.length() - r + 1;
        }
        return ans;
    }
}